src/python/pants/backend/experimental/python/lockfile_metadata.py

```diff
@@ -32,34 +36,47 @@ def from_lockfile(cls, lockfile: bytes) -> LockfileMetadata:
             elif in_metadata_block:
                 metadata_lines.append(line[2:])
 
+        if not metadata_lines:
+            # TODO(#12314): Add a good error.
+            raise InvalidLockfileError("")
+
         try:
             metadata = json.loads(b"\n".join(metadata_lines))
         except json.decoder.JSONDecodeError:
             # TODO(#12314): Add a good error.
-            raise
-
-        T = TypeVar("T")
+            raise InvalidLockfileError("")
 
-        def coerce(t: Callable[[Any], T], key: str) -> T | None:
-            """Gets a value from `metadata`, coercing it to type `t` if not `None`."""
-            v = metadata.get(key, None)
+        def get(key: str) -> Any:
```

Comment:
I think having this function here reduces clarity somewhat -- note that both the `isinstance` check and the `InterpreterConstraints` constructor will fail in a useful way if the variable is `None`, so you could just use `metadata.get(key)` here:


```
hex_digest = metadata.get("requirements_invalidation_digest")
if not isinstance(hex_digest, str):
    raise InvalidLockfileError("")
```

Reply:
Good point! 

I forgot to leave a comment explaining btw why I got rid of coerce. With the upcoming "requirements are compatible, not identical" proposal, the method wasn't flexible enough to convert into list[Requirement].

Reply:
Ah wait, nvm. I remember why I had this helper function: I want the error to be specifically about the key missing. Vs an error for the value being malformed / wrong type. 

Suggestions on a better name for the helper function?

Reply:
`coerce((lambda requirements: [Requirement(i) for i in requirements]), …` works just fine, tbqh :)

Reply:
`get_or_raise` would do fine if you really want to keep the method here.

I suspect, given the work in #12612, we're going to want to just `metadata[key]` natively and have different code for `requirements` (which must be there) vs `InterpreterConstraints`/`Platforms` where 1 and only 1 must be present.

Reply:
> where 1 and only 1 must be present.

Only 1 must be _defined_, but I think we may want both the _present_. I.e. keys vs values. As discussed in DM, I think there's value to seeing `platforms: null` because it gives a cue that the lockfile is platform-agnostic.